### PR TITLE
[draft] post-subsidy security - problem page

### PR DESCRIPTION
**Pull request by VzxPLnHqr**
This is a *very rough* draft for a problem page pertaining to post-subsidy security (issue #29). I do not think it should be merged until it is considerably more readable, less opinionated, etc. 

However, as I started to work on it, it became clear that this notion of "security budget" is not currently very well defined. Or, if it is defined somewhere, it does not seem to be a generally accepted definition. As such, if any of you have definitions or ideas pertaining to a "security budget" that you would like me to include, please comment here or in issue #29.

### CODE REVIEW

**Review by LLFourn — COMMENTED** [review]
hey @VzxPLnHqr. Thanks for writing this up and sorry for the late response. After considering your draft here I am left with this main question: Is this really about the subsidy at all? The security of Bitcoin could fail tomorrow when there still remains a reasonable subsidy. Why should our approach to this problem change when there is a large subsidy vs a small subsidy vs none at all. The presumption is that the security is sufficient now that there is a subsidy but that it might not be after there the subsidy is lower. What if the security is not sufficient right now even? What is sufficient? There is no clert way to answer these questions which probably means that we are notthinking about the question in the right way. Asking the right question so someone can find an answer is of course the main point of bitcoinproblems.org!

So I think there are two questions which are separate. 
1. What should we do if Bitcoin is successfully attacked (loses persistence or liveness). Whether this is post-subsidy or whether fees are high at this time not is not important. As we've identified it probably is important whether this failure is due to SHA256 collisions or not.
2. What do we do if the post-subsidy environment leads to an inconsistent hashrate (i.e. poor hashrate when fees low) which makes an attack more likely. This is important because it's  we can potentially do something about it before an attack happens.

what do you think?

**Comment by VzxPLnHqr** [comment]
@LLFourn No problem on the delay. I have just been letting this marinate in the back of my head for a while now too. 

>Is this really about the subsidy at all? The security of Bitcoin could fail tomorrow when there still remains a reasonable subsidy. Why should our approach to this problem change ...

I think this is a good observation. In my mind, whatever "solution(s)" we come up with should ideally be applicable at all times during the lifecycle of bitcoin, including now.

>So I think there are two questions which are separate.
>
> 1. What should we do if Bitcoin is successfully attacked (loses persistence or liveness). Whether this is post-subsidy or whether fees are high at this time not is not important. As we've identified it probably is important whether this failure is due to SHA256 collisions or not.

Yes, I think a similar way to describe this is: what active measures might we want to have prepared and "waiting in the wings" per se, yet hope we will never have to use them. One of the possible failures is sha256 collisions, but there are of course others, such as a breakdown in secp256k1. We could try to enumerate the various different failure vectors, and that is probably something which should be done anyway. Not sure whether that is the best use of research time currently, but maybe so.

> 2. What do we do if the post-subsidy environment leads to an inconsistent hashrate (i.e. poor hashrate when fees low) which makes an attack more likely. This is important because it's we can potentially do something about it before an attack happens.

This direction is probably more in-line with what I am currently more "interested" in. There may be some considerable design space that is still relatively unexplored. I am not convinced that introducing a small amount of inflation necessarily solves the problem. Some people seem to think so (I think Peter Todd at one point stated that he thought there should be a a small amount of inflation indefinitely?). 

Off the top of my head, if the goal is to perpetually attract future work from miners, then one way that could be done *without inflation* would be to soft-fork in at block X that every block after block X must prove that it is allocating, say, 1000 sats in an anyone-can-spend output (e.g. to be earned by future miners) which is locked for, say, 1 year. The parameters of course could be fine tuned.

Now if, hypothetically, we were to land on a "solution" similar to what I just described above, it still does not really answer the questions about security budget, adequately sizing it, etc. It simply ensures that there will always be some sats remaining to be earned by future miners. Perhaps that assurance alone is enough though.

Anyway, happy to keep thinking on this stuff and hopefully make some progress. I am also not at all tied to whatever draft I originally wrote up for this problem page, so maybe better to start from scratch again.

**Comment by LLFourn** [comment]
Cool. I think we're on the same page.

In terms of keeping the fee pressure more even there was this proposal: https://scalingbitcoin.org/transcript/telaviv2019/elastic-block-caps


> Off the top of my head, if the goal is to perpetually attract future work from miners, then one way that could be done _without inflation_ would be to soft-fork in at block X that every block after block X must prove that it is allocating, say, 1000 sats in an anyone-can-spend output (e.g. to be earned by future miners) which is locked for, say, 1 year. The parameters of course could be fine tuned.

I  still don't understand this. Where are the sats coming from if not inflation?

**Comment by VzxPLnHqr** [comment]
TL;DR: without introducing inflation, the mechanisms available seem to all involve introducing some sort of "pay it forward" game. If such a game is carefully designed, then it could be quite smooth for the majority of the lifetime of the game. It would just become an understood (and soft-fork enforced) convention. Getting the game started would be the rough part, and, then when the game ends it would be rough too. However, the whole point of this is to create a situation and align incentives where the game does not end! Hence, soft-forking in a minimum pay-it-forward bounty per block is perhaps an interesting direction to think about.

> In terms of keeping the fee pressure more even there was this proposal: https://scalingbitcoin.org/transcript/telaviv2019/elastic-block-caps

Thanks for this reference.

> I still don't understand this. 

Apologies if I was unclear. I am simply referring to an idea which would need to be soft-fork enforced by upgraded nodes whereby it is basically a perpetual pay-it-forward game. Something along the lines of:
1. A soft-fork enforced requirement that blocks pay half their fees forward to some time `T` in the future.
2. A soft-fork then would deem blocks invalid if they do not include special output of the form `<T> OP_CHECKLOCKTIMEVERIFY` with an amount greater than or equal to the amount of sats in the usual coinbase fee output).
3. The location of the special output could be defined by the soft fork, similar to how (my understanding of) segwit requires miners to commit the witness data to a specific spot in the block.
4. To cause the game to last (essentially) forever, the soft-fork could require that the duration of the locktime for these pay-it-forward stats grows (slowly, but predictably) as we get further from the genesis of the soft-fork itself.

Item 4 above is what allows for participants (users, future-users, and future-miners) to have confidence that there will always be some sats to be claimed in the future so they should not hesitate to jump in the game now (and stay in!) which is what we want.

> Where are the sats coming from if not inflation?
The sats come from fees paid by spenders and claimed by miners, just like they are today. 

Basically this is just some ramblings on the subject. Probably not super helpful yet!

File: _problems/post-subsidy-security.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+## Impact
+
+<!-- - Try not to repeat the description too much -->
+<!-- - Make it clear what the impact on the big picture of Bitcoin's evolution would be -->
+
+Ultimately a reduced security, if sustained for long enough, manifests as reduced confidence in the future of the network.
+Reduced confidence could trigger a further reduction in security, and a downward spiral ensues. If the 
+network cannot offer an attractive enough reward to those contributing to its security (miners), then this threat, while far
+off right now (2022), could become an existential threat.
```
Comment (ariard) [inline _problems/post-subsidy-security.md:24]: One more idea that could be introduced. Hashrate can be considered as a security offer of which the demand in bitcoins react as an expression of users preferences over a choice of store of value (other cryptocurrencies, gold, real estate, ...) against a given threat and risks model. Higher the hashrate, higher would be the resources and coordination cost encumbered by any reorg entity.

File: _problems/post-subsidy-security.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+1. Sztorc claims this can be solved in [1] via merged-mining and proposes a per-block security budget 
+   `SB` as `SB = P*Q` where `P` is the average transaction price paid in sats/vbyte, and `Q` is the 
+   number of bytes sold, the block size, in bytes/block. 
+   
+2. ...
+   
+ The above are just example definitions and informal possible directions. Yet, if there is to be a formally
+ defined notion of a security budget for the network, then we also should be asking what the appropriate "target"
+ for that budget is. We probably seek some goldilocks scenario here where the amount of security garnered by
+ the network is "not too much," definitely "not too little," but "just right" (in expectation of course).
```
Comment (ariard) [inline _problems/post-subsidy-security.md:52]: I think you could define an adequate security "target" in function of a threat model, where you encompass attacker's available resources and capabilities.

See also recent literature (not parsed yet) seeking for economic limits of Bitcoin: https://ericbudish.org/wp-content/uploads/2018/06/Economic-Limits-Bitcoin-Blockchain-June-2022.pdf

File: _problems/post-subsidy-security.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+   index with a locking script of `<1 year> OP_CSV`. A more complex soft-fork could
+   make the duration a function of the the age of the soft-fork itself.
+
+2. Nothing is broken. Of course there is a reasonable chance that nothing is actually
+   broken and needs fixing here. For example, even in a post-subsidy world where
+   blocks are mostly empty, we can imagine layer-2,3,4... protocols requiring/utilizing
+   recent block headers within their own protocol. Using Lightning as a more concrete
+   example: if everybody is happy with their channels, there is plenty of liquidity,
+   closings are infrequent, etc, then we can easily imagine miners themselves earning
+   fees within channels.
```
Comment (ariard) [inline _problems/post-subsidy-security.md:70]: There is also the possibility that upper layers increase the economic openness of Bitcoin for a large class of low-budget users, of which the sum could paid for a compelling non-subsidy fees.

**Review by ariard — COMMENTED** [review]
> What should we do if Bitcoin is successfully attacked (loses persistence or liveness). Whether this is post-subsidy or whether fees are high at this time not is not important. As we've identified it  probably is important whether this failure is due to SHA256 collisions or not.

+1 on LLoyd's feedback that the problem should be better defined here. I think there are 2 high-level problems with "Post-subsidy security", one is the liveliness security of the network. Beyond cryptographic wreckage, I think pure economic attacks against the network could be considered e.g hashrate is rented with an aim to provoke reorg havoc. The other problem is the subsidy drops in itself, if it would lower the liveliness security of the network, and if something should be done, if any.

**Comment by VzxPLnHqr** [comment]
@ariard Thank you for the review/comments. I am concerned that what we are attempting to tackle in this current page draft here is simply too broad and that the problems should be broken down into more manageable scopes and sub-problems. Hence, this page probably needs a complete rewrite. I will try to get to it eventually, but if someone else wants to take a pass at it, that would be very welcome.

**Comment by VzxPLnHqr** [comment]
@ariard @LLFourn I have not surveyed the "altcoin" landscape much, but I do wonder: are there any existing networks which are already "living" in such a post-subsidy world and, so far at least, seem to be economically viable, surviving on fees alone? 

If there are such examples, then, at least in some ways, they serve as partial counter examples to the claim that "tail emission is necessary."

**Comment by VzxPLnHqr** [comment]
I am not sure if this is the right direction to take it, but I just pushed a commit which removes a lot of the existing text. Given that this is such a broad topic, I figure it may be best to try to initially just publish the problem page and then start refining it as some ideas/proposals become more fully formed in our minds (and links and/or summaries of them are contributed to the page).

**Comment by LLFourn** [comment]
Thanks for making a continued effort in this area @VzxPLnHqr. Sorry for the delay in responding -- I've been on holiday.

I think we need to refocus this page on who our audience is. Our audience is researchers looking for specific problems that they could tackle. I do understand that there is a general concern that fees are not enough to provide "sufficient" security for Bitcoin. I just don't know how to extract it into a problem page in the form this website aims to present them.  You've done a good job trying but TBH your effort have done more to convince me that this broad concern is not really a research topic in itself.

On the other hand I think this discussion has helped us refine two coherent possible problem pages:

1. In the long run, the only "security budget" Bitcoin has is the fees people are willing to pay on transactions. Right now that budget is not being spent optimally because transactions are not broadcast uniformly with respect to days of the week and timezones. What can we do to make the security budget be spent evenly across time?
2. Bitcoin's security assumptions may fail. A 51% can happen regardless of the size of the subsidy or the fees people are paying. What can users do should this eventuate?

I think we should create issues/pages for these two problems instead of this merging this PR. Thoughts?

**Comment by VzxPLnHqr** [comment]
@LLFourn Thanks for your reply. I think you summed it up nicely. Feel free to close this PR in favor of some more specialized issues/pages.

**Comment by ariard** [comment]
> I have not surveyed the "altcoin" landscape much, but I do wonder: are there any existing networks which are already "living" in such a post-subsidy world and, so far at least, seem to be economically viable, surviving on fees alone?

To the best of my knowledge, no. Note to be a compelling data point in my opinion you would need to be competitive in term of offered hashrates, as it's a strong parameter of wider public confidence in Bitcoin. 

It could be interesting to compare the emission schedules of other altcoins, to see how the halvenings there are affecting their corresponding hashrates production, though that's a wide project in itself.

**Comment by VzxPLnHqr** [comment]
@ariard Thanks for your reply. I agree it would be interesting information, but would also be a wide project in itself. 

Perhaps a better testbed would be some sort of (ideally, trustless) PoW sidechain which is not necessarily merge-mined. Here is one idea in that direction:

1. it is already possible to [work-lock](https://github.com/VzxPLnHqr/sig-pow) a utxo such that it is spendable only with sufficient PoW which is verified by the bitcoin network (this can be done today on mainnnet by using `op_size` on an ecdsa signature).
2. imagine a utxo which holds `n` sats, but which has been work-locked and requires an amount `w` of work to unlock those sats
3. it is possible to set `w` to an amount which is more costly to achieve than the reward `n` pays, in which case nobody will even try to unlock it (this is somewhat akin to a post-subsidy bitcoin scenario where the cost to mine exceeds the reward from fees).
4. it is also possible to set `w` to an amount which is too easy (this is somewhat akin to a post-subsidy bitcoin scenario where difficulty has dropped to a point where chain-reorgs are cheap and common).

So, the question is, if we start with a fixed reward `n` to be locked in a utxo, does there exist a procedure for choosing `w` (which must also be chosen ahead of time in order to do the locking) such that cooperation and forward progress (of unlock attempts) prevails? 

One can easily imagine any participants interested in unlocking such a utxo might form their own bitcoin-like protocol which exists for the express purpose of unlocking the utxo. They might even reward each other with their own internal currency where earlier participants receive a greater portion of the `n` sats (when it is finally unlocked). This emulates the subsidy. 

They can transfer their "future sats" among themselves in much the same way as we do with mainchain sats today. They could conceivably even have their own lightning network, and other higher layer protocols.

In fact, and this is one of the primary points I am trying to tease out, there could be more than one group which tries to unlock these same sats. Each group has its own protocol which may or may not be open or otherwise compatible/recognized by the other groups. 

Now, it would not take long for all the groups the realize the conundrum here. They each know they have a risk of putting work in only to see a competing group ultimately make off with the reward. Yet they also know that the other groups are in the same boat. The work from everyone is necessary, otherwise it is unlikely the reward will be unlocked at all. So the groups decide to make their protocols interoperable and (essentially) end up with the bitcoin protocol :-)
